Mathlib/GroupTheory/Perm/Cycle/Basic.lean

```diff
@@ -985,3 +987,154 @@ theorem sum_mul_sum_eq_sum_perm (hσ : σ.IsCycleOn s) (f g : ι → α) :
+
+/-- Restrict a permutation to a Finset containing its support -/
+def subtypePerm_of_support_le (c : Perm α) {s : Finset α}
+    (hcs : c.support ⊆ s) : Equiv.Perm s :=
+  subtypePerm c (isInvariant_of_support_le hcs)
+
+/-- Support of a cycle is nonempty -/
+theorem IsCycle.nonempty_support {g : Perm α} (hg : g.IsCycle) :
+    g.support.Nonempty := by
+  rw [Finset.nonempty_iff_ne_empty, ne_eq, support_eq_empty_iff]
+  exact IsCycle.ne_one hg
+
+/-- Centralizer of a cycle is a power of that cycle on the cycle -/
+theorem IsCycle.commute_iff' {g c : Perm α} (hc : c.IsCycle) :
+    Commute g c ↔
+      ∃ hc' : ∀ x : α, x ∈ c.support ↔ g x ∈ c.support,
+        subtypePerm g hc' ∈ Subgroup.zpowers c.subtypePermOfSupport := by
+  constructor
+  · intro hgc
+    let hgc' := mem_support_iff_of_commute hgc
+    use hgc'
+    obtain ⟨a, ha⟩ := IsCycle.nonempty_support hc
+    obtain ⟨i, hi⟩ := hc.sameCycle (mem_support.mp ha) (mem_support.mp ((hgc' a).mp ha))
+    use i
+    ext ⟨x, hx⟩
+    simp only [subtypePermOfSupport, Subtype.coe_mk, subtypePerm_apply]
+    rw [subtypePerm_apply_zpow_of_mem]
+    obtain ⟨j, rfl⟩ := hc.sameCycle (mem_support.mp ha) (mem_support.mp hx)
+    simp only [← mul_apply, Commute.eq (Commute.zpow_right hgc j)]
+    rw [← zpow_add, add_comm i j, zpow_add]
+    simp only [mul_apply, EmbeddingLike.apply_eq_iff_eq]
+    exact hi
+  · -- converse
+    rintro ⟨hc', h⟩
+    obtain ⟨i, hi⟩ := h
+    ext x
+    simp only [coe_mul, Function.comp_apply]
+    by_cases hx : x ∈ c.support
+    · -- hx : x ∈ c.support
+      suffices hi' : ∀ x ∈ c.support, g x = (c ^ i) x by
+        rw [hi' x hx, hi' (c x) (apply_mem_support.mpr hx)]
+        simp only [← mul_apply, ← zpow_add_one, ← zpow_one_add, add_comm]
+      -- proof of hi'
+      intro x hx
+      let hix := Perm.congr_fun hi ⟨x, hx⟩
+      simp only [← Subtype.coe_inj, subtypePermOfSupport, Subtype.coe_mk, subtypePerm_apply,
+        subtypePerm_apply_zpow_of_mem] at hix
+      exact hix.symm
+    · -- hx : x ∉ c.support
+      rw [not_mem_support.mp hx, eq_comm, ← not_mem_support]
+      contrapose! hx
+      exact (hc' x).mpr hx
+
+/-- A permutation `g` commutes with a cycle `c` if and only if
+  `c.support` is invariant under `g`, and `g` acts on it as a power of `c`. -/
+theorem IsCycle.commute_iff {g c : Perm α} (hc : c.IsCycle) :
+    Commute g c ↔
+      ∃ hc' : ∀ x : α, x ∈ c.support ↔ g x ∈ c.support,
+        ofSubtype (subtypePerm g hc') ∈ Subgroup.zpowers c := by
+  rw [IsCycle.commute_iff' hc]
+  apply exists_congr
+  intro hc'
+  simp only [Subgroup.mem_zpowers_iff]
+  apply exists_congr
+  intro k
+  unfold subtypePermOfSupport
+  rw [subtypePerm_zpow c k]
+  simp only [Equiv.ext_iff, subtypePerm_apply, Subtype.mk.injEq, Subtype.forall]
+  apply forall_congr'
+  intro a
+  by_cases ha : a ∈ c.support
+  · rw [imp_iff_right ha]
+    apply Eq.congr rfl
+    apply symm
+    exact ofSubtype_apply_of_mem _ (by exact ha)
```

Comment:
Ugh, this is annoying, but this works:
```suggestion
  · rw [imp_iff_right ha, ofSubtype_apply_of_mem (p := (· ∈ c.support)) _ ha, subtypePerm_apply]
```
I tried with these lemmas:
```lean
lemma ofSubtype_subtypePerm_of_mem {p : α → Prop} [DecidablePred p] 
    {g : Perm α} {hg : ∀ (x : α), p x ↔ p (g x)}
    {a : α} (ha : p a) : (ofSubtype (g.subtypePerm hg)) a = g a := by
  apply ofSubtype_apply_of_mem
  apply ha

lemma ofSubtype_subtypePerm_of_mem' [DecidableEq α]
    {g c : Perm α} {hg : ∀ (x : α), x ∈ c.support ↔ g x ∈ c.support}
    {a : α} (ha : a ∈ c.support) : (ofSubtype (g.subtypePerm hg)) a = g a := by
  apply ofSubtype_apply_of_mem
  apply ha
```
but `ofSubtype_subtypePerm_of_mem'` is too specialized and `ofSubtype_subtypePerm_of_mem` doesn't work

Reply:
`ofSubtype_subtypePerm_of_mem` works if one gives it the correct argument `hg`. Thanks.